tests/test_common.py

```diff
@@ -0,0 +1,482 @@
+    # # Try signing with an unsupported key type.
+    # # Cheap way to do this: change the listed type of an existing key.
+    # # Another cheap way to do this: change SUPPORTED_KEY_TYPES (but this is
+    # # even more artificial, to the point of probably not being useful)).
+    # # TODO: Consider constructing an artificial Python dict representation of
+    # # a key of a type we don't support. (Not useful?)
+    # key_badtype = copy.deepcopy(keys_pri['primary'])
+    # key_badtype['keytype'] = 'nonsense_type'
+    # with self.assertRaises(uptane.Error):
+    #   common.sign_signable(fresh_ecu_manifest4, [key_badtype],
+    #       datatype='ecu_manifest')
+
+
+
+
+
+  def test_verify_signature_over_metadata(self):
+    """
+    # TODO: Test cases that aren't covered in test_signing() above, for example
+    # expected signature mismatches.
+    This is also tested in other test modules.
+    """
+    pass
+
+
+
+
+
+  def test_canonical_key_funcs(self):
+    """"""
+    # This is also tested by test_primary and test_secondary.
+
+    # Temps for line length:
+    pubkey = keys_pub['secondary']
+    prikey = keys_pri['secondary']
+
+    canonical_key = common.canonical_key_from_pub_and_pri(pubkey, prikey)
+
+    tuf.formats.ANYKEY_SCHEMA.check_match(canonical_key)
+
+    pubkey2 = common.public_key_from_canonical(canonical_key)
+
+    tuf.formats.ANYKEY_SCHEMA.check_match(canonical_key)
+
+    # Break test code if the assumptions we've made for this test to work
+    # correctly are broken.
+    for field in ['keyid', 'keytype', 'keyval']:
+      assert field in pubkey, 'Test code assumption broken.'
+      assert field in prikey, 'Test code assumption broken.'
+    assert 'private' in prikey['keyval'], 'Test code assumption broken.'
+    assert 'public' in pubkey['keyval'], 'Test code assumption broken.'
+    assert 'public' in prikey['keyval'], 'Test code assumption broken.'
+    assert prikey['keyval']['public'] == pubkey['keyval']['public'], \
+        'Test code assumption broken.'
+    assert 'keyid_hash_algorithms' in pubkey2, 'Test code assumption broken.'
+
+    for field in ['keyid', 'keytype', 'keyval', 'keyid_hash_algorithms']:
+      self.assertIn(field, canonical_key)
+      self.assertIn(field, pubkey2)
+
+    for field in ['keyid', 'keytype']:
+      self.assertEqual(canonical_key[field], pubkey[field])
+      self.assertEqual(canonical_key[field], prikey[field])
+      self.assertEqual(canonical_key[field], pubkey2[field])
+
+    for field in ['public', 'private']:
+      self.assertIn(field, canonical_key['keyval'])
+
+    self.assertNotIn('private', pubkey2['keyval'])
+
+
+    self.assertEqual(
+        canonical_key['keyval']['private'], prikey['keyval']['private'])
+    self.assertEqual(
+        canonical_key['keyval']['public'], pubkey['keyval']['public'])
+    self.assertEqual(
+        pubkey2['keyval']['public'], pubkey['keyval']['public'])
+
+
+
+
+
+  def test_create_directory_structure_for_client(self):
```

Comment:
I find it easier to review/verify what is being tested if there were test functions for specific uptane.common functions, instead of one test for multiple functions (like `test_canonical_key_funcs()`).  If you are pressed for time and want to minimize LOC, I guess it's fine.

Reply:
Agreed generally. Maybe here, in particular. In this case, testing each properly would sensibly test them against each other as well. I'm not sure. In any case, I've adjusted the docstrings for the multi-function tests (test_signing() and test_canonical_key_funcs()) to make it easy to see what they test.